src/commons/editor/Comments.tsx

```diff
@@ -0,0 +1,362 @@
+import { CommentAPI } from './useComments';
+import { EventEmitter } from 'events';
+import { some, omit } from 'lodash';
+
+export interface IComment {
+  id: string;
+  isCollapsed: boolean;
+  // TODO: Reference user differently.
+  username: string;
+  profilePic: string;
+  linenum: number;
+  text: string;
+  datetime: number; // if this is infinity, means not submitted yet!
+  // Infinity so it gets sorted to bottom.
+}
+
+export interface CommentsProps {
+  comments: IComment[]; // Ordering guaranteed to be ascending order.
+  commentsBeingEditedRef: React.MutableRefObject<{ [id: string]: IComment }>;
+  APIRef: React.MutableRefObject<CommentAPI>;
+  commentEditChangeEE: EventEmitter;
+}
+
+// =============== STYLES ===============
+// TODO: REMOVE.
+
+const commentsContainerStyles: CSS.Properties = {
+  display: 'grid',
+  gridTemplateColumns: '3em auto',
+  fontFamily: `-apple-system, "BlinkMacSystemFont", "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Open Sans", "Helvetica Neue", "Icons16", sans-serif`
+};
+
+const commentStyles: CSS.Properties = {
+  gridColumn: 2,
+  display: 'grid',
+  gridTemplateColumns: '4em auto',
+  fontSize: '0.8em',
+  backgroundColor: '#253545'
+};
+
+const contentStyles: CSS.Properties = {};
+
+const optionStyles: CSS.Properties = {
+  float: 'right',
+  padding: '0.3em'
+};
+
+const relativeTimeStyles: CSS.Properties = {
+  color: 'lightgray'
+};
+
+const usernameStyles: CSS.Properties = {
+  fontWeight: 'bolder'
+};
+
+const profilePicStyles: CSS.Properties = {
+  width: '3em',
+  height: '3em'
+};
+
+const replyContainerStyles: CSS.Properties = {
+  gridColumn: '1 / 3',
+  padding: '0.5em',
+  backgroundColor: 'grey'
+};
+
+const enterMessageStyles: CSS.Properties = {
+  width: '100%',
+  display: 'block',
+  resize: 'vertical',
+  marginBottom: '0.4em'
+};
+
+const errorTextStyles: CSS.Properties = {
+  color: 'red',
+  padding: '0.2em'
+};
+
+/* Note on interfacing with extra data.
+
+The currentComment will be overwritten by parent refreshing from new data. 
+-Either: Dump in redux store / localstorage.
+-Get a ref to the element from the parent, then request it for data, update and restore it.
+-This WILL cause the focus on the textArea to be lost 
+ because it will be re-rendered by the parent component.
+*/
+
+// Mock function, please replace.
+function sendToServer(comment: IComment) {
+  return new Promise((resolve, reject) => {
+    if (Math.random() < 0.5) {
+      setTimeout(resolve, 1000);
+    } else {
+      setTimeout(() => reject('(Test error message) Some error occured, please try again'), 1000);
+    }
+  });
+}
+
+export default function Comments(props: CommentsProps) {
+  const { comments, commentsBeingEditedRef, APIRef, commentEditChangeEE } = props;
+  // Errors are local.
+  const [errorMsgs, setErrorMsgs] = React.useState({} as { [id: string]: string });
+  const errorMsgsRef = React.useRef(errorMsgs);
+  errorMsgsRef.current = errorMsgs;
```

Comment:
Wait is it necesssary to set `xRef.current = x`? Doesn't React initialise `.current` of the ref to the passed argument `initialValue`?

Reply:
Frankly not sure how it works. E.g. Is the ref created once? Or you need to set current at each step? This is my non familiarity with the library.

Reply:
I suspect it is, at least from what I infer from React's documentation: "`useRef` returns a mutable ref object whose `.current` property is initialized to the passed argument (`initialValue`) - the returned object will persist for the full lifetime of the component."

Well, we can leave it in for now, and remove it later and see if it breaks something.